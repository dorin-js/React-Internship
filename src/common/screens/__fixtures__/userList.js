export const usersList = [
  {
    _created: 1668698852.871516,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1668698852.871524,
    _self_link: 'https://crudapi.co.uk/api/v1/users/fde94e38-a0e8-4244-bbce-65943235bb47',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: 'fde94e38-a0e8-4244-bbce-65943235bb47',
    birth: '2022-11-28',
    email: 'dorin.ilusca@hotmail.com',
    lastname: 'ggg',
    name: '6346366',
  },
  {
    _created: 1668511679.512981,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1668511679.512994,
    _self_link: 'https://crudapi.co.uk/api/v1/users/9d549a7f-f881-47fb-b615-16fd0b81078e',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: '9d549a7f-f881-47fb-b615-16fd0b81078e',
    birth: '2022-09-27',
    email: 'ilusca97@gmail.com',
    lastname: 'ILUSCA',
    name: 'DORIN',
  },
];
