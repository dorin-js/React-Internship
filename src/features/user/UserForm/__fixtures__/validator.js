/* eslint-disable no-useless-escape */
export default {
  firstname: {
    required: {
      value: true,
      message: 'Input is required',
    },
    maxLength: {
      value: 30,
      message: 'The input is too long',
    },
    pattern: {
      value: /^[a-zA-Z]+$/,
      message: 'Name should contain letters only',
    },
  },
  lastname: {
    required: {
      value: true,
      message: 'Input is required',
    },
    maxLength: {
      value: 30,
      message: 'The input is too long',
    },
    pattern: {
      value: /^[a-zA-Z]+$/,
      message: 'Name should contain letters only',
    },
  },
  email: {
    required: {
      value: true,
      message: 'Email is required',
    },
    maxLength: {
      value: 30,
      message: 'The input is too long',
    },
    pattern: {
      value: /^\w+([\.-]?\w+)*@\w+([\.]\w+)(\.\w{2,5})?$/,
      message: 'Incorrect email format',
    },
  },
  birthDate: {
    required: {
      value: true,
      message: 'Birth date is required',
    },
  },
};
