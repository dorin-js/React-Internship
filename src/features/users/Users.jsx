import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import CreateUser from './CreateUser';
import { Modal } from '../../common/components/Modal';
import { UsersTable, UserRow } from './UsersTable';
import { useGetAllUsersQuery } from '../../services/api/apiService';
import { getUserDetails, setDetails } from './usersSlice';

const Users = () => {
  const details = useSelector(getUserDetails);
  const dispatch = useDispatch();
  const {
    data: users,
    isLoading,
    isSuccess,
    isError,
    error,
  } = useGetAllUsersQuery();

  const onUserDetails = (userData) => {
    dispatch(setDetails(userData));
  };

  let content;

  if (isLoading) {
    content = <h3>Loading...</h3>;
  }

  if (isSuccess) {
    content = (
      <>
        <CreateUser />
        <UsersTable>
          {users.map((user) => (
            <UserRow
              key={user.id}
              user={user}
              onUserDetails={onUserDetails}
            />
          ))}
        </UsersTable>
      </>
    );
  }

  if (isError) {
    content = <div>{error.toString()}</div>;
  }

  return (
    <>
      <h2 className="title">Users List</h2>
      {content}
      {details && (
        <Modal onClose={() => onUserDetails(null)}>
          <h5>{details.name}</h5>
          <h5>{details.lastname}</h5>
          <h5>{details.email}</h5>
          <h5>{details.birth}</h5>
        </Modal>
      )}
    </>
  );
};

export default Users;
