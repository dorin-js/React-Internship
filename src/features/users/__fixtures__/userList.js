export const usersList2 = [
  {
    id: 'e512ee0a-0022-4041-ad26-479e425cbaab',
    email: 'jake@mail.com',
    birth: '2022-10-24',
    lastname: 'Sully',
    name: 'Jake',
  },
  {
    id: 'd31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    email: 'jonhndoe@mail.co',
    birth: '2022-11-16',
    lastname: 'Doe',
    name: 'John',
  },
];
export const usersList = [
  {
    _created: 1669148545.821376,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1669148545.821387,
    _self_link: 'https://crudapi.co.uk/api/v1/users/e512ee0a-0022-4041-ad26-479e425cbaab',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: 'e512ee0a-0022-4041-ad26-479e425cbaab',
    id: 'e512ee0a-0022-4041-ad26-479e425cbaab',
    birth: '2022-10-24',
    email: 'jake@mail.com',
    lastname: 'Sully',
    name: 'Jake',
  },
  {
    _created: 1669146430.114581,
    _data_type: 'users',
    _is_deleted: false,
    _modified: 1669371087.72434,
    _self_link: 'https://crudapi.co.uk/api/v1/users/d31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    _user: 'ca1db56c-83bf-4fca-b2e8-6c31fd33f6ff',
    _uuid: 'd31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    birth: '2022-11-16',
    email: 'jonhndoe@mail.co',
    id: 'd31330d2-8d7c-4d63-abcd-bbfa0b7c705f',
    lastname: 'Doe',
    name: 'John',
  },
];
