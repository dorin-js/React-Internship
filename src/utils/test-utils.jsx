import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { PropTypes } from 'prop-types';
import { setupStore } from '../app/store';

export const renderWithProviders = (
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
  } = {},
) => {
  const Wrapper = ({ children }) => (
    <Provider store={store}>{children}</Provider>
  );

  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return { store, ...render(ui, { wrapper: Wrapper }) };
};
