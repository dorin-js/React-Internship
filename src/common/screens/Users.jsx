/* eslint-disable no-unused-vars */
/* eslint-disable no-underscore-dangle */
import React, { useEffect, useState, useCallback } from 'react';
import Error from '../components/Error';
import CreateUser from './CreateUser';
import { UsersTable, UserRow } from '../../features/users/UsersTable';
import Portal from '../components/Portal';
import { Modal } from '../components/Modal';
import { usersApi } from '../services/usersApi/usersApi';
import useGetUsers from '../hooks/useGetUsers';

const Users = () => {
  const [execute, setUsers, { users, loading, error }] = useGetUsers();
  const [details, setDetails] = useState();

  useEffect(() => {
    execute();
    console.log(users);
  }, [execute]);

  const onDoneDelete = (id) => {
    setUsers((prevUsers) => [...prevUsers].filter((user) => user._uuid !== id));
  };

  const onDoneCreate = (items) => {
    setUsers((prevState) => [...items, ...prevState]);
  };

  const onUserDetails = (userData) => {
    setDetails(userData);
  };

  if (loading) {
    return <h3>Loading...</h3>;
  }

  if (error) {
    return <Error message={error} />;
  }

  return (
    <>
      <CreateUser onDoneCreate={onDoneCreate} />
      <UsersTable>
        {users.map((user) => (
          <UserRow
            // eslint-disable-next-line no-underscore-dangle
            key={user._uuid}
            user={user}
            onDoneDelete={onDoneDelete}
            onUserDetails={onUserDetails}
          />
        ))}
      </UsersTable>
      {details && (
        <Portal>
          <Modal onClose={() => setDetails(null)}>
            <h5>{details.name}</h5>
            <h5>{details.lastname}</h5>
            <h5>{details.email}</h5>
            <h5>{details.birth}</h5>
          </Modal>
        </Portal>
      )}
    </>
  );
};

export default Users;
