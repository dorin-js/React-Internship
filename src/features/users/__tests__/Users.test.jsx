import React from 'react';
import {
  act, fireEvent, render, waitFor, screen,
} from '@testing-library/react';
import Users from '../Users';
import { usersList } from '../__fixtures__/userList';
import { renderWithProviders } from '../../../utils/test-utils';
import {
  useGetAllUsersQuery,
  useDeleteUserMutation,
  useUpdateUserMutation,
  useCreateNewUserMutation
} from '../../../services/api/apiService';

jest.mock('../../../services/api/apiService', () => ({
  useGetAllUsersQuery: jest.fn(),
  useDeleteUserMutation: jest.fn(),
  useUpdateUserMutation: jest.fn(),
  useCreateNewUserMutation: jest.fn(),
}));


describe('Users Screen', () => {
  beforeEach(() => {
    useGetAllUsersQuery.mockImplementation(() => ({}));
    useDeleteUserMutation.mockImplementation(() => ([]));;
    useCreateNewUserMutation.mockImplementation(() => ([]));
    useUpdateUserMutation.mockImplementation(() => ([]))
  });

  it('should fetch the users data by calling the hook useGetAllUsersQuery', () => {
    renderWithProviders(<Users />);
    expect(useGetAllUsersQuery).toHaveBeenCalled();
  });

  describe('while loading', () => {
    it('should render "Loading..."', () => {
      useGetAllUsersQuery.mockImplementation(() => ({
        isLoading: true,
      }));

      const { getByText } = renderWithProviders(<Users />);

      const loaderElement = getByText(/loading/i);
      expect(loaderElement).toBeInTheDocument();
    })
  })

  describe('if there is an error', () => {
    it('should render an error message', () => {
      useGetAllUsersQuery.mockImplementation(() => ({
        isError: true,
        error: { error: 'TypeError: Failed to fetch' },
      }));

      const { getByText } = renderWithProviders(<Users />);

      const errorElement = getByText(/TypeError: Failed to fetch/i);
      expect(errorElement).toBeInTheDocument();
    })
  })

  describe('with data', () => {
    beforeEach(() => {
      const createNewUser = jest.fn();
      const updateUser = jest.fn();
      const deleteUser = jest.fn();
      useUpdateUserMutation.mockImplementation(() => ([
        updateUser,
        { isLoading: false, isError: false, error: null }
      ]));
      useDeleteUserMutation.mockImplementation(() => ([
        deleteUser,
        { isLoading: false, isError: false, error: null }
      ]));
      useCreateNewUserMutation.mockImplementation(() => ([
        createNewUser,
        { isSucces: true, isLoading: false, isError: false, error: null }
      ]));
    })

    it('renders correctly', async () => {
      const { baseElement } = renderWithProviders(<Users />);
      expect(baseElement).toMatchSnapshot();
    });

    it('should render UsersTable and CreateUser components', () => {
      useGetAllUsersQuery.mockImplementation(() => ({
        data: usersList,
      }));

      const { getByText, getByRole } = renderWithProviders(<Users />);
      screen.debug();
      const usersTable = getByRole('table');
      const rowElement = getByText('John');
      expect(usersTable).toBeInTheDocument();
      expect(rowElement).toBeInTheDocument();
    });
  })
});