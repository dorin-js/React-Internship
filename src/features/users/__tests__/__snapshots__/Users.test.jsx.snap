// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Screen renders correctly 1`] = `undefined`;
